invokeai/app/services/images/images_default.py

```diff
@@ -24,15 +29,220 @@
+                image_data.height = height
+                image_name = self.__invoker.services.names.create_image_name()
+                image_data.image_name = image_name
+                # if image_data.image.size[0] < 5000:  # Fail condition: fail if width or height is too large
+                #     raise Exception("Intentional failure for testing: Image size too large")
+                self.__invoker.services.image_records.save(
+                    image_name=image_data.image_name,
+                    image_origin=image_data.image_origin,
+                    image_category=image_data.image_category,
+                    width=image_data.width,
+                    height=image_data.height,
+                    has_workflow=image_data.workflow is not None,
+                    is_intermediate=image_data.is_intermediate,
+                    metadata=image_data.metadata,
+                    node_id=image_data.node_id,
+                    session_id=image_data.session_id,
+                    starred=image_data.starred,
+                )
+
+                if image_data.board_id is not None:
+                    self.__invoker.services.board_image_records.add_image_to_board(board_id=image_data.board_id, image_name=image_data.image_name)
+
+                self.__invoker.services.image_files.save(
+                    image_name=image_data.image_name, image=image_data.image, metadata=image_data.metadata, workflow=image_data.workflow
+                )
+
+                image_dto = self.get_dto(image_data.image_name)
+                self._on_changed(image_dto)
+
+                with progress_lock:
+                    processed_counter += 1 
+
+                return image_dto
+            except ImageRecordSaveException:
+                self.__invoker.services.logger.error("Failed to save image record")
+                raise
+            except ImageFileSaveException:
+                self.__invoker.services.logger.error("Failed to save image file")
+                raise
+            except Exception as e:
+                self.__invoker.services.logger.error(f"Problem processing and saving image: {str(e)}")
+                raise e
+        
+        # Determine the number of available CPU cores
+        num_cores = os.cpu_count() or 1
+        num_workers = max(1, num_cores - 1)
+
+        images_DTOs = []
+
+        # Initialize tqdm progress bar
+        pbar = tqdm(total=total_images, desc="Processing Images", unit="images", colour="green")
+
+        with ThreadPoolExecutor(max_workers=num_workers) as executor:
+            futures = [executor.submit(process_and_save_image, image) for image in upload_data_list]
+            for future in as_completed(futures):
+                try:
+                    image_dto = future.result()
+                    images_DTOs.append(image_dto)
+                    pbar.update(1)  # Update progress bar
+
+                    progress_percentage = (processed_counter / total_images) * 100
```

Comment:
We already have `processed` and `total`, do we need percentage in this payload? It's trivial to derive it clientside